src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -753,21 +764,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
             {
                 var index = command.Index;
                 var options = mergedCommands.Options[index];
-                Task indexTask = null;
+                var taskResult = new ClusterTransactionResult
+                {
+                    GeneratedResult = mergedCommands.Replies[index],
+                };
                 if (options.WaitForIndexesTimeout != null)
                 {
-                    indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
+                    var indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
                         options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
                             mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
```

Comment:
I don't think that this is needed at all.
Why do you _care_ about this separately? You don't care about the task result here any longer, do you?
That is handled by the batch handler

Reply:
`ClusterTransactionWaiter.CreateTask` will create a task if it doesn't exist, and if it exists, it will take it instead of creating a new one. 

If we remove the `ClusterTransactionWaiter.CreateTask` from here, it can create some race in the case of 'wait for index'.

between this line:
https://github.com/ravendb/ravendb/blob/1a4a13a3f6176e1a6e08be66fe0dfc1519bad8e9/src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs#L68
and this line:
https://github.com/ravendb/ravendb/blob/1a4a13a3f6176e1a6e08be66fe0dfc1519bad8e9/src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs#L71C12-L71C12

Suppose between those lines the cluster transaction is already completed in the database (the `LastCompletedClusterTransactionIndex` is equal to or greater than the cluster transaction index) but the indexing wasn't completed yet.
In that case, it won't wait for the indexing.

Reply:
Then move the `CreateClusterTransactionTask` _before_ the `SendToLeaderAsync`, no?

Reply:
You cant, because you need the pass index for creating the task